sonar-auth-saml-plugin/src/main/java/org/sonarsource/auth/saml/SamlIdentityProvider.java

```diff
@@ -66,11 +93,100 @@ public boolean allowsUsersToSignUp() {
+    }
   }
 
   @Override
   public void callback(CallbackContext context) {
-    // TODO
+    Auth auth = newAuth(initSettings(null), context.getRequest(), context.getResponse());
+    processResponse(auth);
+    csrfVerifier.verifyState(context.getRequest(), context.getResponse());
+
+    LOGGER.trace("Name ID : {}", auth.getNameId());
+    checkAuthentication(auth);
+
+    LOGGER.trace("Attributes received : {}", auth.getAttributes());
+    String login = getNonNullFirstAttribute(auth, samlSettings.getUserLogin());
+    UserIdentity.Builder userIdentityBuilder = UserIdentity.builder()
+      .setLogin(login)
+      .setProviderLogin(login)
+      .setName(getNonNullFirstAttribute(auth, samlSettings.getUserName()));
+    samlSettings.getUserEmail().ifPresent(
+      email -> userIdentityBuilder.setEmail(getFirstAttribute(auth, email)));
+    samlSettings.getGroupName().ifPresent(
+      group -> userIdentityBuilder.setGroups(getGroups(auth, group)));
+    context.authenticate(userIdentityBuilder.build());
+    context.redirectToRequestedPage();
+  }
+
+  private static Auth newAuth(Saml2Settings saml2Settings, HttpServletRequest request, HttpServletResponse response) {
+    try {
+      return new Auth(saml2Settings, request, response);
+    } catch (SettingsException e) {
+      throw new IllegalStateException("Fail to create Auth", e);
+    }
+  }
+
+  private static void processResponse(Auth auth) {
+    try {
+      auth.processResponse();
+    } catch (Exception e) {
+      throw new IllegalStateException("Fail to process response", e);
+    }
+  }
+
+  private static void checkAuthentication(Auth auth) {
+    List<String> errors = auth.getErrors();
+    if (auth.isAuthenticated() && errors.isEmpty()) {
+      return;
+    }
+    String errorReason = auth.getLastErrorReason();
+    throw new UnauthorizedException(errorReason != null && !errorReason.isEmpty() ? errorReason : "Unknown error reason");
+  }
+
+  private static String getNonNullFirstAttribute(Auth auth, String key) {
+    String attribute = getFirstAttribute(auth, key);
+    requireNonNull(attribute, String.format("%s is missing", key));
+    return attribute;
+  }
+
+  @CheckForNull
+  private static String getFirstAttribute(Auth auth, String key) {
+    Collection<String> attribute = auth.getAttribute(key);
+    if (attribute == null || attribute.isEmpty()) {
+      return null;
+    }
+    return attribute.iterator().next();
+  }
+
+  private static Set<String> getGroups(Auth auth, String groupAttribute) {
+    Collection<String> attribute = auth.getAttribute(groupAttribute);
+    if (attribute == null || attribute.isEmpty()) {
+      return emptySet();
+    }
+    return new HashSet<>(attribute);
+  }
+
+  private Saml2Settings initSettings(@Nullable String callbackUrl) {
+    Map<String, Object> samlData = new HashMap<>();
+    // TODO strict mode is unfortunately not compatible with HTTPS configuration on reverse proxy => https://jira.sonarsource.com/browse/SQAUTHSAML-8
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.auth.saml%3Asonar-auth-saml&pullRequest=4&issues=AWTqn0ZS7yorTq7tcpUE&open=AWTqn0ZS7yorTq7tcpUE)

[ISSUE_KEY]: AWTqn0ZS7yorTq7tcpUE